src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs

```diff
@@ -247,19 +250,34 @@ public async void WorkerError(WorkerErrorEvent workerError)
             {
                 _logger.LogDebug("Handling WorkerErrorEvent for runtime:{runtime}, workerId:{workerId}", workerError.Language, workerError.WorkerId);
                 _languageWorkerErrors.Add(workerError.Exception);
-                bool isPreInitializedChannel = _webHostLanguageWorkerChannelManager.ShutdownChannelIfExists(workerError.Language, workerError.WorkerId);
-                if (!isPreInitializedChannel)
+                await DisposeAndRestartWorkerChannel(workerError.Language, workerError.WorkerId);
+            }
+        }
+
+        public async void WorkerRestart(WorkerRestartEvent workerRestart)
+        {
+            if (!_disposing)
+            {
+                _logger.LogDebug("Handling WorkerRestartEvent for runtime:{runtime}, workerId:{workerId}", workerRestart.Language, workerRestart.WorkerId);
+                await DisposeAndRestartWorkerChannel(workerRestart.Language, workerRestart.WorkerId);
+            }
+        }
+
+        private async Task DisposeAndRestartWorkerChannel(string runtime, string workerId)
+        {
+            bool isPreInitializedChannel = _webHostLanguageWorkerChannelManager.ShutdownChannelIfExists(runtime, workerId);
+            if (!isPreInitializedChannel)
+            {
+                _logger.LogDebug("Disposing errored channel for workerId: {channelId}, for runtime:{language}", workerId, runtime);
```

Comment:
Nit: remove `errored` from the log as this method is now a helper to dispose error or restart event